src/webgpu/api/validation/render_pass_descriptor.spec.ts

```diff
@@ -192,6 +199,57 @@ g.test('attachments_must_match_whether_they_are_used_for_color_or_depth_stencil'
   }
 });
 
+g.test('attachments_must_match_fragment_outputs_base_type')
+  .paramsSubcasesOnly([
+    { format: 'rgba8unorm', _success: true }, // valid
+    { format: 'rgba8unorm-srgb', _success: false },
+    { format: 'r8unorm', _success: false },
+    { format: 'bgra8unorm', _success: false },
+    { format: 'r16float', _success: false },
+    { format: 'r32uint', _success: false },
+    { format: 'rgba32float', _success: false },
+  ])
+  .fn(async t => {
+    const { format, _success } = t.params;
+
+    const pipeline = t.device.createRenderPipeline({
+      vertex: {
+        module: t.device.createShaderModule({
+          code: `
+        [[stage(vertex)]] fn main(
+          [[builtin(vertex_index)]] VertexIndex : u32
+          ) -> [[builtin(position)]] vec4<f32> {
+            return vec4<f32>(0.0, 0.0, 0.0, 1.0);
+          }
+          `,
+        }),
+        entryPoint: 'main',
+      },
+      fragment: {
+        module: t.device.createShaderModule({
+          code: `
+          [[stage(fragment)]] fn main() -> [[location(0)]] vec4<f32> {
+            return vec4<f32>(1.0, 0.0, 0.0, 1.0);
+          }
+          `,
+        }),
+        entryPoint: 'main',
+        targets: [{ format: 'rgba8unorm' }],
```

Comment:
Since the format is hardcoded in the pipeline here this test is testing the compatibility between the pipeline and the render pass. (Is that the intent?)
I think this validation is covered in `api,validation,attachment_compatibility:render_pass_or_bundle_and_pipeline,color_format`

However there's another piece of validation, which is compatibility between the fragment shader itself (the `vec4<f32>` in this test) and the `.fragments.targets`. (This validation occurs inside createRenderPipeline.)
Unfortunately the spec is not very specific about this validation yet:
https://gpuweb.github.io/gpuweb/#abstract-opdef-validating-gpufragmentstate
> The pipeline output type must be compatible with colorState.format.

But I'm pretty sure the validation is:
- The scalar type (f32, i32, or u32) must match the [sample type of the format](https://gpuweb.github.io/gpuweb/#plain-color-formats).
- The size of the vec (e.g. `f32`, `vec2<f32>`, `vec3<f32>`, `vec4<f32>`) doesn't matter. Extra components are discarded and missing components are filled to `0,0,1`.

But we should verify that (by testing, probably).

I don't _think_ we have any tests for that yet. Possible places:
- `api,validation,createRenderPipeline` doesn't seem to have it though it _does_ generate shaders with matching primitive types
- `api,validation,layout_shader_compat` is only for compatibility between the shader and the pipeline layout, not other parts of the pipeline
- `api,validation,attachment_compatibility` is only for compatibility between pipeline, pass, and bundle.

so I'd put that in either createRenderPipeline or a new file.

Reply:
> (by testing, probably)

Validation testing will help here (it will make sure we don't pass anything to the drivers that the debug layers don't like), but really we'll need operation testing to make sure the results come through properly.

Maybe you can try to see if we need to add an operation test plan for that.